File: backend/app/services/audit_service.py
from typing import Any, Optional
from sqlalchemy.orm import Session
from app.models.audit_log import AuditLog

class AuditService:
    @staticmethod
    def log(
        db: Session,
        *,
        user_id: Optional[Any] = None,
        username: Optional[str] = None,
        action: str,
        target_type: Optional[str] = None,
        target_id: Optional[str] = None,
        details: Optional[dict] = None,
        ip_address: Optional[str] = None
    ) -> AuditLog:
        db_log = AuditLog(
            user_id=user_id,
            username=username,
            action=action,
            target_type=target_type,
            target_id=target_id,
            details=details,
            ip_address=ip_address
        )
        db.add(db_log)
        db.commit()
        db.refresh(db_log)
        return db_log

audit_service = AuditService()

File: frontend/src/services/auditService.ts
import api from "@/services/api";
import { PaginatedResponse } from "@/types/pagination";

export interface AuditLog {
  id: string;
  user_id: string | null;
  username: string | null;
  action: string;
  target_type: string | null;
  target_id: string | null;
  details: any;
  ip_address: string | null;
  created_at: string;
}

export const auditService = {
  getLogs: async (page = 1, limit = 8) => {
    const skip = (page - 1) * limit;
    const response = await api.get<PaginatedResponse<AuditLog>>("/audit-logs/", { params: { skip, limit } });
    return response.data;
  },

  exportLogsPDF: async () => {
    const response = await api.get("/audit-logs/export/pdf", {
      responseType: 'blob'
    });
    const url = window.URL.createObjectURL(new Blob([response.data]));
    const link = document.createElement('a');
    link.href = url;
    link.setAttribute('download', 'audit_logs_export.pdf');
    document.body.appendChild(link);
    link.click();
    link.remove();
  },
};
